waspc/e2e-test/Main.hs

```diff
@@ -0,0 +1,54 @@
+import Control.Monad (unless)
+import System.Directory (getCurrentDirectory)
+import System.FilePath (joinPath, takeFileName, (</>))
+import System.Info
+import Test.Tasty (TestTree, defaultMain, testGroup)
+import Util
+  ( runGoldenTest,
+    waspCliNew,
+    waspCliNewAndCompile,
+    waspCliNewAndCompileAndMigrate,
+  )
+
+-- TODO: Put `cwd` into ReaderT Monad?
+
+-- TODO: Switch over from FilePath to StrongPath eventually and use helper like so:
+--
+-- data ProjectRoot
+-- getProjectRootPath :: IO (Path Abs (Dir ProjectRoot))
+-- getProjectRootPath = getCurrentDirectory
+
+type GoldenTest = FilePath -> IO TestTree
+
+main :: IO ()
+main = do
+  if os == "mingw32"
+    then putStrLn "Skipping end-to-end tests on Windows due to tests using *nix-only commands"
+    else do
+      -- Stack/Cabal launch `stack test` from root of project, so this should always be some absolute path to waspc
+      cwd <- getCurrentDirectory
+      unless (takeFileName cwd == "waspc") (error "Expecting test process to be invoked from waspc dir")
+      defaultMain =<< tests cwd
+
+tests :: GoldenTest
+tests cwd =
+  testGroup "All Golden Dir Tests"
+    <$> sequence
+      [ waspNew cwd,
+        waspCompile cwd,
+        waspMigrate cwd
+      ]
+
+waspNew :: GoldenTest
+waspNew cwd =
+  runGoldenTest cwd "waspNew" waspCliNew
+
+waspCompile :: GoldenTest
+waspCompile cwd =
+  runGoldenTest cwd "waspCompile" waspCliNewAndCompile
+
+waspMigrate :: GoldenTest
+waspMigrate cwd =
+  runGoldenTest cwd "waspMigrate" (waspCliNewAndCompileAndMigrate cwd (cwd </> joinPath ["e2e-test", "goldens", "overlays", "entity.txt"]))
```

Comment:
Here `waspCliNewAndCompileAndMigrate` takes a lot more arguments, and I'm not sure why. I think if the test was inlined and flatter (not a separate function in another file) it'd be easier for me to understand what's going on.

(with tests in my opinion many rules of factoring change; simple flat tests often are easier to understand and maintain than test code split into functions)

Reply:
That is the overlay file, but yeah arguably putting too much logic into the commands and less into the tests is not ideal. The pro to extracting is the commands can be used in other, larger tests. We can chat more on it?